Account deletion and leave group logic

TomRomeo opened the pull request:
This PR implements the requested feature of account deletion to comply with GDPR / DSGVO (#61)

The endpoint is `DELETE` `/api/user` 

With account deletion, the following data is also ereased:
- revlog entries
- user-card bindings
- user-group-role bindings

> **Important**
> With the current implementation, if a user is part of a group, the group does **not** get deleted. Therefore, home groups will remain and groups can end up without any admins.

How would you propose we solve that problem?
Should 'empty' groups be deleted? What about non-empty groups where no admin would remain?

darmiel commented on backend/pkg/model/cardside.go:
```diff
@@ -8,7 +8,7 @@ import (
 	ID                 string `gorm:"primaryKey"`
 	CardID             string `gorm:"not null"`
-	Card               Card
+        Card               Card `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
```
what happened to the format?

darmiel commented on backend/pkg/model/groupUserRole.go:
```diff
@@ -14,6 +14,6 @@ type GroupUserRole struct {
 	RoleType RoleType
-	Group    Group `gorm:"constraint:OnUpdate:CASCADE"`
-	User     User  `gorm:"constraint:OnUpdate:CASCADE"`
+        Group    Group `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
```
go fmt plz

darmiel commented on backend/pkg/model/revlog.go:
```diff
@@ -10,7 +10,7 @@ type Revlog struct {
 	Card   Card
 	UserID string `gorm:"not null"`
-	User   User
+        User   User `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
```
🧐

darmiel commented on backend/pkg/model/userCardBinding.go:
```diff
@@ -8,6 +8,7 @@ import (
 type UserCardBinding struct {
 	ID           string `gorm:"primaryKey"`
 	UserID       string `gorm:"not null"`
+        User        User    `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
```
🤨

darmiel commented on backend/services/user/handler/user.go:
```diff
@@ -56,6 +56,22 @@ func (e *User) Register(ctx context.Context, req *pb.RegisterRequest, rsp *pb.Na
 
+func (e *User) DeleteUser(ctx context.Context, req *pb.UserID, rsp *pb.SuccessResponse) error {
+	logger.Infof("Received User.Delete request: %v", req)
+        user, err := e.store.FindUserByID(req.UserID)
+        if err != nil {
+            logger.Errorf("Could not find user with userid: %s: %s", user.ID, err)
+            return err
+        }
+        err = e.store.DeleteUser(user)
+        if err != nil {
+            logger.Errorf("An error occured while trying to delete user with ID '%s': %s", user.ID, err)
+            return err
+        }
```
😡

isn't your GoLand set to automatically format?

TomRomeo replied:
wrote with nvim™

memeToasty replied:
![aCtUaLlY](https://media.tenor.com/eRD89_uiAvsAAAAC/nerd-well-actually.gif)

darmiel replied:
isn't your nvim set to automatically format?

darmiel reviewed (changes requested):
👑 (only formatting)

Julian702 commented:
From my point of view, we should also delete the Home Group including all decks to ensure that no user data is left.

With empty public groups, the question is whether an external user can learn decks from the group. As far as I remember we had agreed that you don't have to join a group to learn (reference and not copy) the public decks of the group. The empty group could be deleted, but not the decks in the group that are referenced by other users. However, since no one has the rights to edit these decks anymore, it would be a possibility to replace the references with copies. 
Private empty groups can be deleted in any case, since nobody has access to them.

In this context, let's discuss again and define what private and public means, to cover all cases. For example, can a public group also have private decks?
My suggestion would be to specify the attribute private and public only for decks, as the current solution is confusing and duplicated. Private decks can then be deleted and references to public decks can be replaced with copies as described above before deletion.
For groups we should then save whether you can join them. I could imagine the values "Anyone can join", "Request to join" or "Closed". In any case, an empty group would then be deleted.

We should also think about keeping non-user related data for a short time to be able to restore it if needed or if this is handled by our backup strategy in the future.

In the case that there is no admin left, from my point of view users would have to create a new group and copy decks, as I can't think of a reliable strategy to designate a new admin.

Julian702 commented:
> I don't think we should keep the user data as a backup, if the user decides to delete the account, it should be irreversible. Because a backup would lead to new problems.

As I described above, the backup was not about user-related data but about groups and decks, for example. I could imagine deleting the associations, but not the entries in the database itself, so that they could be restored.

> For the case of the deletion of an admin account of a group I would suggest to promote the user with the highest role to be the admin. It is an easier procedure for the moment as we don't have the "copy" feature yet.

What happens then if several users have the same role? Will one of them be promoted? Will everyone be promoted? I think both solutions are problematic, because users without write access could even delete the group for everyone after promotion. I think the simplest solution is not to promote anyone at first, so that no user can lose data that way. Afterwards, we can address the copy feature.

TomRomeo commented:
I started work on #56 as well, since this issue covers the same group logic.
The current approach would be:
1. User leaves group
2. if no user remain -> delete group
3. if user remain but no admin remains -> also delete group

I think this might be the most straight forward solution. A group admin can transfer ownership before leaving the group but just setting a random user to admin is not a good solution, I think.

As for handling of private / public decks / copies, I think we have to further specify this behavior in another meeting / issue / PR, because as Kevin mentioned, this is not implemented yet at all.

Julian702 commented:
> 2. if no user remain -> delete group

This is currently only possible with private groups until we can automatically replace references to public decks.

> 3. if user remain but no admin remains -> also delete group

This is not possible at all, since the members would lose their decks. In this case, the group must remain to give them the possibility to continue learning or to transfer the decks to their home group in the future.

> A group admin can transfer ownership before leaving the group but just setting a random user to admin is not a good solution, I think.

I agree, I think the best option would be to force the last admin to either delete the group or make another group member the admin before he is allowed to leave the group.

TomRomeo commented:
> This is currently only possible with private groups until we can automatically replace references to public decks.

> This is not possible at all, since the members would lose their decks. In this case, the group must remain to give them the possibility to continue learning or to transfer the decks to their home group in the future.

I think it is more sensible to implement it this way first (deleting groups as I described above) and later change the logic to account for referenced decks and team members, since this requires a lot of logic that is not implemented yet and at the current rate would lead to most groups not being deleted and dangling without any admins, which is quite ugly in my opinion.

Julian702 commented:
As we discussed in our last Weekly, here is a brief summary of the simple solution we have agreed on for now. 

When a user deletes their account, the following steps must be completed.

1. Delete all account information, including the user's home group and all group memberships.

2. Delete all private decks in empty groups (currently all decks in private groups, as we do not differentiate here).

3. Delete all groups without members and without decks.

This simple logic prevents that members of a group lose their decks, because the last admin deletes his account and the group is removed as well.

Regarding leaving a group described in #56, we agreed that we force the last admin to promote another group member or delete the group. This has to be implemented mostly in the frontend, as the backend only has to return an error here. I will create a new issue for the frontend implementation and link it here.

Julian702 commented:
I encountered a few problems during testing:

- When I try to delete my account I get the following response (500):
   ```no existing user with email```
- When I try to leave my non-empty group as non-admin or as last admin I get the following response (200):
   ```
   {
      success: true
   }
   ```
   However, I am still in the group and can repeat the request several times with the same answer. Besides, I think that the status code 200 should be sufficient and a body is not needed here.

TomRomeo commented:
You are right, I indeed missed this case in the leave group logic. I have now added it.
Of course, we would not have this problem if we would just simplify the group logic to deleting the group once the last admin leaves but oh well. 😭

I have also changed the returned response to align with already existing responses.

Julian702 commented:
Thanks for fixing the bug described above. Unfortunately I found another two problems during testing:

- response from DELETE /api/user still not matching our project default
- when deleting the last admin account in a private group, the group ist deleted although there are other members in this group

I have not tested all possible scenarios at this point. Please check the logic for deleting decks and groups after account deletion I described in #61. Could you also remove the code smell detected by SonarCloud?

Julian702 commented:
When I started adapting the frontend, I noticed that the two endpoints feel a bit inconsistent. 

DELETE `/api/user`
DELETE `/api/groups/G-LwvbfUvs/members/leave`

We should keep a consistent scheme here, considering the promotion and demotion endpoints to be implemented next. What's your opinion?

TomRomeo commented:
I finally came around to fix the logic and refactor the endpoints according to your comments. Could you take a look at this code again? @Julian702 @4KevR

4KevR commented on backend/services/frontend/handler/frontend.go:
```diff
@@ -177,6 +189,20 @@ func (e *Frontend) GetUserHandler(c *fiber.Ctx) error {
+	rspDeleteUser, err := e.userService.DeleteUser(c.Context(), &pbUser.UserID{
+		UserID: userID,
+	})
+	if err != nil {
+		return err
+	} else if !rspDeleteUser.Success {
+		return helper.NewMicroNotSuccessfulResponseErr(helper.FrontendServiceID)
+	} else {
+		return c.SendStatus(200)
+	}
```
```suggestion
	return c.SendStatus(200)
```

darmiel replied:
```suggestion
	if err != nil {
		return err
	}
	if !rspDeleteUser.Success {
		return helper.NewMicroNotSuccessfulResponseErr(helper.FrontendServiceID)
	}
	return c.SendStatus(200)
```

4KevR commented on backend/services/frontend/handler/frontend.go:
```diff
@@ -177,6 +189,20 @@ func (e *Frontend) GetUserHandler(c *fiber.Ctx) error {
+	})
+	if err != nil {
+		return err
+	} else if !rspDeleteUser.Success {
+		return helper.NewMicroNotSuccessfulResponseErr(helper.FrontendServiceID)
```
```suggestion
	if !rspDeleteUser.Success {
	      return helper.NewMicroNotSuccessfulResponseErr(helper.FrontendServiceID)
```

4KevR commented on backend/services/frontend/handler/frontend.go:
```diff
@@ -226,6 +252,24 @@ func (e *Frontend) CreateGroupHandler(c *fiber.Ctx) error {
+			GroupID: groupID,
+		},
+	)
+	if err != nil {
+		return err
+	} else if !rspLeaveGroup.Success {
+		return helper.NewMicroNotSuccessfulResponseErr(helper.FrontendServiceID)
+	}
```
s.a.

darmiel replied:
```suggestion
	if err != nil {
		return err
	}
	if !rspLeaveGroup.Success {
		return helper.NewMicroNotSuccessfulResponseErr(helper.FrontendServiceID)
	}
```

darmiel commented on backend/services/collaboration/handler/collaboration.go:
```diff
@@ -465,3 +468,36 @@ func (e *Collaboration) FindGroupByID(_ context.Context, req *pb.GroupRequest, r
+	if err != nil {
+		return err
+	}
+	if len(remainingAdmins) == 1 && remainingAdmins[0].UserID == req.UserID {
+		return helper.NewMicroCantLeaveAsLastAdminErr(helper.CollaborationServiceID)
+	} else {
+		return e.LeaveGroup(ctx, req, rsp)
+	}
```
```suggestion
	if len(remainingAdmins) == 1 && remainingAdmins[0].UserID == req.UserID {
		return helper.NewMicroCantLeaveAsLastAdminErr(helper.CollaborationServiceID)
	}
	return e.LeaveGroup(ctx, req, rsp)
```

4KevR commented on backend/services/user/handler/user.go:
```diff
@@ -57,6 +57,45 @@ func (e *User) Register(ctx context.Context, req *pb.RegisterRequest, rsp *pb.Na
 }
 
+func (e *User) VerifyUserExists(ctx context.Context, req *pb.VerificationRequest, rsp *pb.SuccessResponse) error {
+	usr, err := e.store.FindUserByEmail(req.UserEmail)
```
maybe `user`?

darmiel commented on backend/services/collaboration/handler/collaboration.go:
```diff
@@ -465,3 +468,36 @@ func (e *Collaboration) FindGroupByID(_ context.Context, req *pb.GroupRequest, r
+		e.store.DeleteGroup(group)
+	}
+	rsp.Success = true
+	logger.Infof("User %s left group %s. ", req.UserID, req.GroupID)
```
```suggestion
	logger.Infof("User %s left group %s.", req.UserID, req.GroupID)
```

4KevR reviewed (changes requested):
Looks good, see comments for minor improvements

darmiel commented on backend/store/store.go:
```diff
@@ -40,6 +41,7 @@ type CollaborationStore interface {
 	DeleteGroup(group *model.Group) error
 	GetGroupUserRole(userID string, groupID string) (model.RoleType, error)
 	GetGroupMemberRoles(groupID string) ([]model.GroupUserRole, error)
+	GetGroupAdmins(groupID string) ([]model.GroupUserRole, error)
```
```suggestion
	FindGroupMemberRoles(groupID string) ([]model.GroupUserRole, error)
	FindGroupAdmins(groupID string) ([]model.GroupUserRole, error)
```

TomRomeo replied:
As discussed with @Julian702 I will not change the naming here as it has been used before and is not completely written in one style. Furthermore, the functions are implemented at many points already.

darmiel commented on backend/services/collaboration/handler/collaboration.go:
```diff
@@ -465,3 +468,36 @@ func (e *Collaboration) FindGroupByID(_ context.Context, req *pb.GroupRequest, r
+		return err
+	}
+	if len(groupUsers) > 1 || group.GroupType == model.Public {
+		e.store.RemoveUserFromGroup(req.UserID, req.GroupID)
```
What about the error?

darmiel commented on backend/services/collaboration/handler/collaboration.go:
```diff
@@ -465,3 +468,36 @@ func (e *Collaboration) FindGroupByID(_ context.Context, req *pb.GroupRequest, r
+	if len(groupUsers) > 1 || group.GroupType == model.Public {
+		e.store.RemoveUserFromGroup(req.UserID, req.GroupID)
+	} else if group.GroupType == model.Public {
+		e.store.DeleteGroup(group)
```
This error also waits to be handled :(

darmiel commented on backend/services/collaboration/handler/collaboration.go:
```diff
@@ -465,3 +468,36 @@ func (e *Collaboration) FindGroupByID(_ context.Context, req *pb.GroupRequest, r
 }
+
+func (e *Collaboration) LeaveGroupSafe(ctx context.Context, req *pb.GroupRequest, rsp *pb.SuccessResponse) error {
+	logger.Infof("Received Collaboration.LeaveGroupSafe request: %v", req)
```
tip: use `%+v` instead of `%v` to display the keys in the struct

```go
{ Test 86 nil }
```
```go
{ Name:Test Age:86 Attributes:nil }
```

darmiel commented on backend/services/collaboration/handler/collaboration.go:
```diff
@@ -465,3 +468,39 @@ func (e *Collaboration) FindGroupByID(_ context.Context, req *pb.GroupRequest, r
+		if err = e.store.RemoveUserFromGroup(req.UserID, req.GroupID); err != nil {
+			return err
+		}
+	} else if group.GroupType == model.Public {
```
this code is never executed (`group.GroupType == model.Public` already covered this check above) - you probably just mixed up `||` and `&&`

Julian702 commented on backend/kioku-api.postman_collection.json:
```diff
@@ -1007,6 +1030,37 @@
+				"method": "DELETE",
+				"header": [],
+				"url": {
+					"raw": "127.0.0.1:3000/api/groups/:groupID/members/leave",
```
```suggestion
					"raw": "127.0.0.1:3000/api/groups/:groupID/members",
```

Julian702 reviewed (approved):
Looks good now, please just update the postman collection
